sdk/identity/identity/src/credentials/defaultAzureCredential.ts

```diff
@@ -11,149 +11,158 @@ import {
       ...options,
-    });
+      clientId: managedIdentityClientId,
+    };
+
+    return new ManagedIdentityCredential(managedIdentityClientOptions);
   }
+
+  // TODO: I'll need some help here, because I have a hunch we can return a UnavailableDefaultCredential here but I am not sure
+  return new ManagedIdentityCredential(options);
 }
-export class DefaultAzureCliCredential extends AzureCliCredential {
-  constructor(options?: DefaultAzureCredentialOptions) {
-    super({
-      processTimeoutInMs: options?.processTimeoutInMs,
+
+function createDefaultWorkloadIdentityCredential(
+  options?: DefaultAzureCredentialOptions,
+): TokenCredential {
+  const managedIdentityClientId =
+    (options as DefaultAzureCredentialClientIdOptions)?.managedIdentityClientId ??
+    process.env.AZURE_CLIENT_ID;
+  const workloadIdentityClientId =
+    (options as DefaultAzureCredentialClientIdOptions)?.workloadIdentityClientId ??
+    managedIdentityClientId;
+  const workloadFile = process.env.AZURE_FEDERATED_TOKEN_FILE;
+  const tenantId = options?.tenantId ?? process.env.AZURE_TENANT_ID;
+  if (workloadFile && workloadIdentityClientId) {
+    const workloadIdentityCredentialOptions: WorkloadIdentityCredentialOptions = {
+      ...options,
+      tenantId,
+      clientId: workloadIdentityClientId,
+      tokenFilePath: workloadFile,
+    };
+    return new WorkloadIdentityCredential(workloadIdentityCredentialOptions);
+  }
+  if (tenantId) {
+    const workloadIdentityClientTenantOptions: WorkloadIdentityCredentialOptions = {
       ...options,
-    });
+      tenantId,
+    };
+    return new WorkloadIdentityCredential(workloadIdentityClientTenantOptions);
   }
+
+  // This is a smell to me, and I believe can also be replaced with a UnavailableDefaultCredential
+  return new WorkloadIdentityCredential(options as WorkloadIdentityCredentialOptions);
 }
 
-export class DefaultAzurePowershellCredential extends AzurePowerShellCredential {
-  constructor(options?: DefaultAzureCredentialOptions) {
-    super({
-      processTimeoutInMs: options?.processTimeoutInMs,
-      ...options,
-    });
+function createDefaultAzureDeveloperCliCredential(
+  options: DefaultAzureCredentialOptions = {},
+): TokenCredential {
+  const processTimeoutInMs = options.processTimeoutInMs;
+  return new AzureDeveloperCliCredential({ processTimeoutInMs, ...options });
+}
+
+function createDefaultAzureCliCredential(
+  options: DefaultAzureCredentialOptions = {},
+): TokenCredential {
+  const processTimeoutInMs = options.processTimeoutInMs;
+  return new AzureCliCredential({ processTimeoutInMs, ...options });
+}
+
+function createDefaultAzurePowershellCredential(
+  options: DefaultAzureCredentialOptions = {},
+): TokenCredential {
+  const processTimeoutInMs = options.processTimeoutInMs;
+  return new AzurePowerShellCredential({ processTimeoutInMs, ...options });
+}
+
+function createEnvironmentCredential(
```

Comment:
You can start to see the benefits here, and I kept it as small as possible just as an example, but if we are missing data we know the credential needs let's just not create the credential.

Reply:
@KarishmaGhiya @mpodwysocki @xirzec any preference between the two options listed here? Or any questions I can help answer? 

I want to clean this up, fix the tests, and publish, but curious what you think so far - would this help us at all?

Reply:
initially I was leaning towards undefined as it's cleaner but I *do* kinda like being able to have a message about what was missing if you find the lack of a credential surprising... but maybe as long as that is logged with the appropriate verbosity it's not needed?

Reply:
> initially I was leaning towards undefined as it's cleaner but I _do_ kinda like being able to have a message about what was missing if you find the lack of a credential surprising... but maybe as long as that is logged with the appropriate verbosity it's not needed?

Oh we require to throw and log the reasoning with the message under the Credential Unavailable Error that gets thrown as part of the AggregateErrors of the chained token credential.